### PR TITLE
Rewrite clientside and towerside arduino codebases

### PR DESCRIPTION
**Pull request by BluCodeGH**
All new code in `src/towerside`, `src/clientside` and `src/common` is commented and ready for review. I'm not super happy with the clientside architecture but we need this code for static fire which is next weekend.

---
This change is [Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35)

### CODE REVIEW

**Review by Joe-Joe-Joe-Joe — COMMENTED** [review]
Most of the actual code is going over my head, but the comments are well done and I can follow what's happening.

**Review by Joe-Joe-Joe-Joe — APPROVED** [review]

**Review by QuantumManiac — APPROVED** [review]
just this one thing. If that's not actually an issue then LGTM.

> Reviewed 95 of 105 files at r1, 11 of 11 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-:-N1kdmGT003h-z0GYYM3:b1n2ahi)* status: all files reviewed, 1 unresolved discussion (waiting on @BluCodeGH and @KavinSatheeskumar)

___
*[`src/clientside/clientside.ino` line 58 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1kY189-uUCHYnctJCa:-N1kY189-uUCHYnctJCb:b-aftaep) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/175ab9731a0120ee4bf96ce2bf51cdbdd28af79b/src/clientside/clientside.ino#L58)):*
> ```cpp
>       // Light up the switch LEDs if we are armed
>       for (uint8_t i = 0; i < sizeof(Pinout::MISSILE_LEDS) / sizeof(Pinout::MISSILE_LEDS[0]); i++) {
>         digitalWrite(Pinout::MISSILE_LEDS[i], !armed);
> ```

is `!armed` correct here if we want to enable LEDs when armed? Or is it active low or something of the sorts?

**Review by BluCodeGH — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35)* status: 105 of 106 files reviewed, 1 unresolved discussion (waiting on @KavinSatheeskumar and @QuantumManiac)

___
*[`src/clientside/clientside.ino` line 58 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1kY189-uUCHYnctJCa:-N1lT9pj0XWWJe6FrhBL:b-sokfnv) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/175ab9731a0120ee4bf96ce2bf51cdbdd28af79b/src/clientside/clientside.ino#L58)):*
<details><summary><i>Previously, QuantumManiac (Chamath Wijesekera) wrote…</i></summary><blockquote>

is `!armed` correct here if we want to enable LEDs when armed? Or is it active low or something of the sorts?
</blockquote></details>

It's active low yeah. Added a comment.

File: src/towerside/hardware.hpp [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#ifndef HARDWARE_H
+#define HARDWARE_H
+
+namespace Hardware {
+
+void setup();
+
+}
+
```
Comment (ZTeper) [inline src/towerside/hardware.hpp:9]: empty function again

File: src/towerside/sensors.hpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        rolling_avg{new uint16_t[rolling_avg_width]} {
+      for (uint8_t i = 0; i < rolling_avg_width; i++) {
+        rolling_avg[i] = 0;
+      }
+      pinMode(pin, INPUT);
+      digitalWrite(pin, false); // turn off any pullup
+    }
+    uint16_t get_value() override {
+      // analog input values are in multiples of ~5 mV (technically 5/1024), so multiply by 5
+      return rolling_sum / rolling_avg_width * 5 * m_num / m_den + b;
```
Comment (ZTeper) [inline src/towerside/sensors.hpp:78]: Please use `()` to clarify the order of operations

**Review by ZTeper — APPROVED** [review]

**Review by BluCodeGH — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35)* status: 101 of 106 files reviewed, 10 unresolved discussions (waiting on @KavinSatheeskumar, @QuantumManiac, and @ZTeper)

___
*[`src/clientside/channels.hpp` line 34 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpEg-r2-34:-N1qLxZQ68OJXaWkReBs:b-vwj7tx) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/175ab9731a0120ee4bf96ce2bf51cdbdd28af79b/src/clientside/channels.hpp#L34)):*
<details><summary><i>Previously, ZTeper (Zachary Teper) wrote…</i></summary><blockquote>

Please define a constant or a macro for numbers like 3, 4, 999 instead of magic numbers
</blockquote></details>

This is a totally valid point that we should definitely fix, but fixing it properly (dynamic field lengths) would mean a pretty big overhaul of the clientside display code. That code is pretty shaky right now and needs and overhaul anyways, in the interest of getting this in I'd rather add it to the existing clickup task https://app.clickup.com/t/29gcgpe.
___
*[`src/clientside/clientside.ino` line 58 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpEh-r2-58:-N1qMaM_E5cif_6sOnbo:b3atlic) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/175ab9731a0120ee4bf96ce2bf51cdbdd28af79b/src/clientside/clientside.ino#L58)):*
<details><summary><i>Previously, ZTeper (Zachary Teper) wrote…</i></summary><blockquote>

why `!armed` ? That would only light up LEDs if we are NOT armed... unless they are active low....
</blockquote></details>

Yeah active low, added a comment.
___
*[`src/clientside/clientside.ino` line 70 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpEh-r2-70:-N1qMdRUB8MXAbwAm6Gs:b7mtcmd) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/175ab9731a0120ee4bf96ce2bf51cdbdd28af79b/src/clientside/clientside.ino#L70)):*
<details><summary><i>Previously, ZTeper (Zachary Teper) wrote…</i></summary><blockquote>

random empty function why ???
</blockquote></details>

Arduino mandates it, added a comment.
___
*[`src/clientside/daq.cpp` line 20 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpEl-r2-20:-N1qNR0I-zapfl6JIpKj:b-896fix) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/175ab9731a0120ee4bf96ce2bf51cdbdd28af79b/src/clientside/daq.cpp#L20)):*
<details><summary><i>Previously, ZTeper (Zachary Teper) wrote…</i></summary><blockquote>

Please define constants/macros for this magic number
</blockquote></details>

Done.
___
*[`src/common/shared_types.hpp` line 34 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpF2-r2-34:-N1qNT3S5CYkLYEmqnXs:bcmw4tp) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/175ab9731a0120ee4bf96ce2bf51cdbdd28af79b/src/common/shared_types.hpp#L34)):*
<details><summary><i>Previously, ZTeper (Zachary Teper) wrote…</i></summary><blockquote>

Use a ternary operator here so that the function only returns 1 or 0
</blockquote></details>

I think the function returning `bool` does this already but added anyways.
___
*[`src/common/communication/decoder.hpp` line 42 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpEu-r2-42:-N1qNqbj3oxGRXdJf8Cc:b-s7yu1x) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/175ab9731a0120ee4bf96ce2bf51cdbdd28af79b/src/common/communication/decoder.hpp#L42)):*
<details><summary><i>Previously, ZTeper (Zachary Teper) wrote…</i></summary><blockquote>

This is a really long function. Perhaps break it up into multiple smaller functions?
</blockquote></details>

Yeah it's ugly but it's basically checking a bunch of sequential cases so I don't see a clean way to break it up.
___
*[`src/common/communication/receiver.hpp` line 32 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpEw-r2-32:-N1qNz1O7Vfe7vrYzIN6:bydlr9h) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/175ab9731a0120ee4bf96ce2bf51cdbdd28af79b/src/common/communication/receiver.hpp#L32)):*
<details><summary><i>Previously, ZTeper (Zachary Teper) wrote…</i></summary><blockquote>

Use of `new` without `delete`. Potential memory leak
</blockquote></details>

Not an issue, one MessageReceiver is instantiated at the start of the program and then never touched. You'll see a lot of naked allocations when setting stuff up since using pointers for polymorphism has way better ergonomics than references (where you need to keep a bunch of global variables around) but the end result in terms of allocated memory is the same.
___
*[`src/towerside/hardware.hpp` line 9 at r3](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpFJ-r3-9:-N1qP4Sn6RUJ8lwskW-d:b1hpd38) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/907095e3e9bf3479b424860113c4aa2e22d6cc52/src/towerside/hardware.hpp#L9)):*
<details><summary><i>Previously, ZTeper (Zachary Teper) wrote…</i></summary><blockquote>

empty function again
</blockquote></details>

This is a .hpp file.
___
*[`src/towerside/sensors.hpp` line 78 at r3](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpFL-r3-78:-N1qPCL16cRMEKZkPBu5:b5fx8ya) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/907095e3e9bf3479b424860113c4aa2e22d6cc52/src/towerside/sensors.hpp#L78)):*
<details><summary><i>Previously, ZTeper (Zachary Teper) wrote…</i></summary><blockquote>

Please use `()` to clarify the order of operations
</blockquote></details>

Precedence between multiplication and division is always left-to-right, which is the intended order (multiply as much as we can without overflowing). I don't think `(((((a / b) * c) * d) / e)` is nearly as readable.

**Review by Artem-Sotnikov — COMMENTED** [review]
Looks great! Some of it seems little overkill but it's very well-structured and I don't see any glaring issues 

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-:-N1q_swq19lsBkqlY5Tq:b9t1ay1)* status: 101 of 106 files reviewed, 13 unresolved discussions (waiting on @BluCodeGH, @KavinSatheeskumar, @QuantumManiac, and @ZTeper)

___
*[`src/clientside/channels.hpp` line 34 at r2](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpEg-r2-34:-N1qXWli4kobQXHgpRci:b-w53sdz) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/175ab9731a0120ee4bf96ce2bf51cdbdd28af79b/src/clientside/channels.hpp#L34)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

This is a totally valid point that we should definitely fix, but fixing it properly (dynamic field lengths) would mean a pretty big overhaul of the clientside display code. That code is pretty shaky right now and needs and overhaul anyways, in the interest of getting this in I'd rather add it to the existing clickup task https://app.clickup.com/t/29gcgpe.
</blockquote></details>

Yeah I think the comments address it well enough and it's good for now.
___
*[`src/clientside/clientside.ino` line 46 at r4](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1qY2NB0adfLdXucp-u:-N1qY2NB0adfLdXucp-v:bz2dbc) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/dcbcfe9eeda5dcd7802b639e97d764330a982a2b/src/clientside/clientside.ino#L46)):*
> ```cpp
>   while (true) {
>     // Call the tick() methods on anything that inherits from Tickable
>     Tickable::trigger_tick();
> ```

I think this Tickable thing is a little overkill and opaque rather than calling everything that needs to be ticked (which can just be a function). But it is versatile and certainly fine for this version.
___
*[`src/clientside/clientside.ino` line 57 at r4](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1qTeR0DZAyPTPkDbF_:-N1qTeR0DZAyPTPkDbFa:bkd5eat) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/dcbcfe9eeda5dcd7802b639e97d764330a982a2b/src/clientside/clientside.ino#L57)):*
> ```cpp
>       }
>       // Light up the switch LEDs if we are armed
>       for (uint8_t i = 0; i < sizeof(Pinout::MISSILE_LEDS) / sizeof(Pinout::MISSILE_LEDS[0]); i++) {
> ```

Why are we dividing by the first element of the array here? 
___
*[`src/common/communication/receiver.hpp` line 49 at r4](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1qbnjEAhxgZy_TpqkS:-N1qbnjEAhxgZy_TpqkT:b-qkarcv) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/dcbcfe9eeda5dcd7802b639e97d764330a982a2b/src/common/communication/receiver.hpp#L49)):*
> ```cpp
>     }
>     // Force the handlers to handle a specific message (eg when transitioning to safe state)
>     void force(const T &msg) {
> ```

I like this functionality, but something like an interrupt might work better for such an application. After all, we're running this on a microcontroller.
___
*[`src/towerside/sensors.hpp` line 78 at r3](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35#-N1KOz-qa8p5gtnJCpFL-r3-78:-N1q_3P96XNUBTNSeQpF:b56ab1p) ([raw file](https://github.com/waterloo-rocketry/rlcs/blob/907095e3e9bf3479b424860113c4aa2e22d6cc52/src/towerside/sensors.hpp#L78)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

Precedence between multiplication and division is always left-to-right, which is the intended order (multiply as much as we can without overflowing). I don't think `(((((a / b) * c) * d) / e)` is nearly as readable.
</blockquote></details>

I think maybe splitting this operation up into two lines will make it more readable.

**Review by KavinSatheeskumar — APPROVED** [review]
> Reviewed 30 of 105 files at r1, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/rlcs/35)* status: all files reviewed, 13 unresolved discussions (waiting on @BluCodeGH, @QuantumManiac, and @ZTeper)

**Pull request merged**